pyopia/pipeline.py

```diff
@@ -82,11 +104,18 @@ def run(self):
             stats (DataFrame): stats DataFrame of particle statistics
         '''
 
-        timestamp, imraw = self.steps['load']()
-        imc = self.steps['imageprep'](imraw, self.common)
-        stats = self.steps['statextract'](timestamp, imc, self.cl)
+        self.common['steps_string'] = steps_to_string(self.common['steps'])
+
+        blacklist_steps = ['common', 'classifier']
+        # [self.steps[s]() for s in self.steps if not start.__contains__(s)]
+        for s in self.common['steps']:
+            if blacklist_steps.__contains__(s):
+                continue
+
+            print('calling: ', str(type(self.common['steps'][s])), ' with: ', list(self.common.keys()))
+            self.common = self.common['steps'][s](self.common)
 
-        self.steps['output'](stats, steps_to_string(self.steps))
+        stats = self.common['stats']
 
         return stats
```

Comment:
Maybe you should simply return 'common', as you are breaking the convention that the user can have the input/outputs they like, and forcing them to "end up at" comon['stats'].

Reply:
fixed :)